### PR TITLE
Misc fixes

### PR DESCRIPTION
**Pull request by Osyotr**
Towards #4600

Each commit can be reviewed separately.

### CODE REVIEW

File: base/base_tests/optional_lock_guard_tests.cpp
Comment (biodranik) [inline base/base_tests/optional_lock_guard_tests.cpp:1]: OptionalLockGuard is not used, right?

File: base/beam.hpp [hunk 1]
```diff
@@ -53,7 +53,15 @@ class Beam
     }
 
     if (m_entries.size() == m_capacity)
+    {
+      if (it == std::prev(m_entries.end()))
+      {
+        m_entries.back() = Entry(key, value);
```
Comment (biodranik) [inline base/beam.hpp:59]: Why is it needed?
CC @vng

File: coding/writer.hpp [hunk 2]
```diff
@@ -61,8 +61,11 @@ template <typename ContainerT> class MemWriter : public Writer
 
   void Write(void const * p, size_t size) override
   {
+    if (size == 0)
+        return;
```
Comment (biodranik) [inline coding/writer.hpp:65]:
```suggestion
      return;
```

File: coding/writer.hpp [hunk 3]
```diff
@@ -61,8 +61,11 @@ template <typename ContainerT> class MemWriter : public Writer
...
+
     auto freeSize = static_cast<intptr_t>(m_Data.size() - m_Pos);
```
Comment (biodranik) [inline coding/writer.hpp:67]: Should size and pos be casted to intptr before the subtraction?

File: qt/mainwindow.cpp [hunk 4]
```diff
@@ -222,9 +222,13 @@ MainWindow::MainWindow(Framework & framework,
   RoutingSettings::LoadSession(m_pDrawWidget->GetFramework());
 }
 
-#if defined(Q_WS_WIN)
-bool MainWindow::winEvent(MSG * msg, long * result)
+#if defined(OMIM_OS_WINDOWS)
+bool MainWindow::nativeEvent(QByteArray const & eventType, void * message, qintptr * result)
 {
+  (void)eventType;
+  (void)result;
```
Comment (biodranik) [inline qt/mainwindow.cpp:229]: Why are these needed?

File: transit/transit_types.hpp [hunk 5]
```diff
@@ -164,7 +164,7 @@ class Stop
   FeatureIdentifiers m_featureIdentifiers;
   TransferId m_transferId = kInvalidTransferId;
   std::vector<LineId> m_lineIds;
-  m2::PointD m_point;
+  m2::PointD m_point = m2::PointD(0.0, 0.0);
```
Comment (biodranik) [inline transit/transit_types.hpp:167]:
```suggestion
  m2::PointD m_point {0.0, 0.0};
```

Comment (biodranik) [reply transit/transit_types.hpp:167]: Why is this initialization necessary?

**Review by biodranik — COMMENTED** [review]
Thanks for the great job! A few comments.

File: base/base_tests/CMakeLists.txt [hunk 6]
```diff
@@ -44,12 +43,16 @@ set(SRC
   thread_safe_queue_tests.cpp
   threaded_list_test.cpp
   threads_test.cpp
-  timegm_test.cpp
   timer_test.cpp
   uni_string_dfa_test.cpp
   visitor_tests.cpp
 )
 
+# Windows does not have timegm function
```
[the remark this answers is not in the document]
Comment (Osyotr) [reply base/base_tests/CMakeLists.txt:51]: Hmm how about we remove this test completely?
https://github.com/organicmaps/organicmaps/commit/0764876ead28d001d9a642f6f15974d38159d4d5#diff-b57d4356b92827b1be2417f54c34998c2cd7461986ef3f1568929ec11af3a7deR15-R16
https://github.com/organicmaps/organicmaps/commit/0764876ead28d001d9a642f6f15974d38159d4d5#diff-424a2c879f2c3da46de6df2bd99e92076aa576f7ba39d74c0207df99e8b77d65R55-R59

File: base/base_tests/optional_lock_guard_tests.cpp
Comment (Osyotr) [reply base/base_tests/optional_lock_guard_tests.cpp:1]: Yes.
Note: I've mentioned it in commit message.

File: base/beam.hpp (hunk 1 above)
Comment (Osyotr) [reply base/beam.hpp:59]: > `pop_back` invalidates `it` if `it == std::prev(m_entries.end())`

Compared to what you did in https://github.com/organicmaps/organicmaps/commit/0764876ead28d001d9a642f6f15974d38159d4d5#diff-a871d373cd08c23fae73306f90a71b35c35e85850ba3ad068ef6bfc61e656914R57, this code does not allow `m_entries` to grow beyond `m_capacity`.

File: qt/mainwindow.cpp (hunk 4 above)
Comment (Osyotr) [reply qt/mainwindow.cpp:229]: Silencing warnings about unreferenced parameters. Let me know if unnamed arguments are more preferred.

File: transit/transit_types.hpp (hunk 5 above)
Comment (Osyotr) [reply transit/transit_types.hpp:167]: Inside `Transit_StopSerialization` test, `routing::transit::Stop` object is created, serialized and deserialized. Since this field is uninitialized, some garbage value was written and then read back. Would be unnoticed but `PointUToPointD` does some bounds checking so the deserialized point does not equal to the original.
image

File: coding/writer.hpp (hunk 3 above)
Comment (Osyotr) [reply coding/writer.hpp:67]: Not needed - `m_Data.size()` is always greater or equal than `m_Pos`.

File: base/beam.hpp (hunk 1 above)
Comment (vng) [reply base/beam.hpp:59]: Did some checker show the error/warning here? Or you found it by yourself?

Actually, I can't say for sure that the previous code was well-formed .. If the iterator points to the last element pop_back invalidates the element, but the iterator remains _good_ for the follow-up insertion (at least I can't imagine some other vector-iterator implementation).

New code is more safe for sure.

File: coding/coding_tests/bit_streams_test.cpp [hunk 7]
```diff
@@ -19,19 +19,19 @@ namespace
 {
 UNIT_TEST(BitStreams_Smoke)
 {
-  uniform_int_distribution<uint8_t> randomBytesDistribution(0, 255);
+  uniform_int_distribution<uint32_t> randomBytesDistribution(0, 255);
```
Comment (vng) [inline coding/coding_tests/bit_streams_test.cpp:22]: Please, show the compiler output here, what is wrong with uint8_t.

File: coding/writer.hpp (hunk 3 above)
Comment (vng) [reply coding/writer.hpp:67]: - Hm, if so, what is the reason in ```if (freeSize < 0)``` logic? We can do Seek in any position according to the code.
- It is not reproducible in our situation, but the current code is ill-formed in a very corner case: https://coliru.stacked-crooked.com/a/0ffdb84c8e972040

File: coding/writer.hpp (hunk 2 above)
Comment (vng) [reply coding/writer.hpp:65]: Honestly saying not needed here, IMO.

File: drape/stipple_pen_resource.hpp [hunk 8]
```diff
@@ -22,7 +22,7 @@ uint32_t constexpr kMaxStipplePenLength = 512;  /// @todo Should be equal with k
 // Based on ./data/patterns.txt, the most of patterns have 2 entries (4 entries for triangles pattern).
 using PenPatternT = buffer_vector<uint16_t, 2>;
 
-inline constexpr uint16_t PatternFloat2Pixel(double d)
+inline uint16_t PatternFloat2Pixel(double d)
```
Comment (vng) [inline drape/stipple_pen_resource.hpp:25]: Well, yes, constexpr here is since C++23

File: geometry/clipping.cpp [hunk 9]
```diff
@@ -86,7 +86,7 @@ void ClipTriangleByRect(m2::RectD const & rect, m2::PointD const & p1, m2::Point
     return;
   }
 
-  double constexpr kEps = 1e-8;
+  static constexpr double kEps = 1e-8;
```
Comment (vng) [inline geometry/clipping.cpp:89]: Why it is better?

File: transit/transit_types.hpp (hunk 5 above)
Comment (vng) [reply transit/transit_types.hpp:167]: Well, yes formally, but who cares if the test checks ```TEST(!stop.IsValid(), (stop));```
Some tool shows an error here or some runtime diagnostics?

**Comment by biodranik** [comment]
@vng many of these issues were discovered by the visual studio when I built the windows and windows2 branches.

File: base/beam.hpp (hunk 1 above)
Comment (Osyotr) [reply base/beam.hpp:59]: > Did some checker show the error/warning here?

MSVC STL detected it. IIRC it was `Beam_Smoke` test.

> If the iterator points to the last element pop_back invalidates the element, but the iterator remains good for the follow-up insertion 

From [cppreference](https://en.cppreference.com/w/cpp/container/vector/pop_back):
> Iterators (including the end() iterator) and references to the last element are invalidated.

My understanding is that "invalid" means "invalid even for insertion".

File: geometry/clipping.cpp (hunk 9 above)
Comment (Osyotr) [reply geometry/clipping.cpp:89]: This variable should be static because it's used without explicit capture in lambda below.

File: coding/writer.hpp (hunk 3 above)
Comment (vng) [reply coding/writer.hpp:67]: Of course, no need to modify this logic, because size and m_Pos are fit into intptr_t.

File: transit/transit_types.hpp (hunk 5 above)
Comment (Osyotr) [reply transit/transit_types.hpp:167]: >  Some tool shows an error here or some runtime diagnostics?

The latter. The test also checks `obj.IsEqualForTesting(deserializedObj)` inside `TestSerialization`/`TestCommonSerialization`.
https://github.com/organicmaps/organicmaps/blob/c1de096229597715914cda0f03050a2093416981/transit/transit_tests/transit_test.cpp#L39

File: coding/coding_tests/bit_streams_test.cpp (hunk 7 above)
Comment (Osyotr) [reply coding/coding_tests/bit_streams_test.cpp:22]: It is included in the commit message:
> error C2338: static_assert failed: 'invalid template argument for uniform_int_distribution: N4950 [rand.req.genl]/1.5 requires one of short, int, long, long long, unsigned short, unsigned int, unsigned long, or unsigned long long'

**Comment by vng** [comment]
LGTM, I'd prefer to remove
```
if (size == 0)
  return;
```
This case is _predict false_ almost always :) Even if size == 0, everything should work ok.

**Comment by Osyotr** [comment]
> UPD: Ok, for sure can rewrite:
> 
> ```
> - &m_Data[static_cast<size_t>(m_Pos)]
> + m_Data.data() + m_Pos
> ```
`memcpy` requires valid pointers even if size is 0...

File: transit/transit_types.hpp (hunk 5 above)
Comment (biodranik) [reply transit/transit_types.hpp:167]: 1. Is the issue limited to tests only?
2. Why 0,0 is a correct initializer? Is a better fix possible?
3. Does it make sense to use some kInvalidPoint check instead of forcing it to zeroes?

File: qt/mainwindow.cpp (hunk 4 above)
Comment (biodranik) [reply qt/mainwindow.cpp:229]: Aren't they referenced below?

I personally prefer `/*message*/` instead, to save vertical lines space.

File: base/base_tests/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (biodranik) [reply base/base_tests/CMakeLists.txt:51]: These links do not work on Github for some reason.

Can `timegm` be replaced by [_mkgmtime](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/mkgmtime-mkgmtime32-mkgmtime64) in tests?
```
#define timegm _mkgmtime
``` 
so we have a working test?

File: coding/writer.hpp (hunk 2 above)
Comment (biodranik) [reply coding/writer.hpp:65]: Is it better to replace by ASSERT?

**Comment by vng** [comment]
@Osyotr 
Ok, my variant to simplify all here:
```
  void Write(void const * p, size_t size) override
  {
    size_t const newSize = m_Pos + size;
    if (m_Data.size() < newSize)
    {
      // Assume the same alloc strategy as in std::vector.
      // +1 to avoid zero checks
      m_Data.resize(std::max(newSize, m_Data.capacity() * 2 + 1));
    }

    memcpy(m_Data.data() + m_Pos, p, size);

    m_Pos += size;
  }

private:
  ContainerT & m_Data;
  size_t m_Pos;
```

File: transit/transit_types.hpp (hunk 5 above)
Comment (Osyotr) [reply transit/transit_types.hpp:167]: > 1. Is the issue limited to tests only?

Yeah, for some reason default constructor is used only in tests (`Transit_StopSerialization`).

> 2. Why 0,0 is a correct initializer? Is a better fix possible?
> 3. Does it make sense to use some kInvalidPoint check instead of forcing it to zeroes?

Any value in bounds [-180;180] would work. See `PointDToPointU` for details. Considering that `Transit_StopSerialization` doesn't fail for linux users I assume `m_point={0.0,0.0}` is what's already happening.

**Comment by Osyotr** [comment]
> m_Data.resize(std::max(newSize, m_Data.capacity() * 2));

`BitStreams_T1` is failing due to this.

File: drape/stipple_pen_resource.hpp (hunk 8 above)
Comment (biodranik) [reply drape/stipple_pen_resource.hpp:25]: Did the generated code change for modern arm64 compilers? If it became slower, it may be better to put constexpr under an ifdef.

**Review by vng — APPROVED** [review]

File: drape/stipple_pen_resource.hpp (hunk 8 above)
Comment (Osyotr) [reply drape/stipple_pen_resource.hpp:25]: There are no usages of this function in constexpr context.
